File: to_be_renaimed/lib/models/meal_plan.dart
import 'package:to_be_renaimed/models/enums.dart';
import 'package:to_be_renaimed/models/ingredient.dart';
import 'package:to_be_renaimed/models/recipe.dart';

class WeeklyMealPlan {
  final int id;
  final DateTime startDate;
  final DateTime endDate;
  final int userId;

  
  List<DailyMealPlan> dailyPlans = [];

  WeeklyMealPlan({
    required this.id,
    required this.startDate,
    required this.endDate,
    required this.userId,
    List<DailyMealPlan>? dailyPlans,
  }) : dailyPlans = dailyPlans ?? [];

  factory WeeklyMealPlan.fromJson(Map<String, dynamic> json) {
    return WeeklyMealPlan(
      id: json['wmp_id'],
      startDate: DateTime.parse(json['wmp_start']),
      endDate: DateTime.parse(json['wmp_end']),
      userId: json['wmp_usr_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'wmp_id': id,
      'wmp_start': startDate.toIso8601String().split('T')[0],
      'wmp_end': endDate.toIso8601String().split('T')[0],
      'wmp_usr_id': userId,
    };
  }

  
  bool containsDate(DateTime date) {
    final dayOnly = DateTime(date.year, date.month, date.day);
    return !dayOnly.isBefore(startDate) && !dayOnly.isAfter(endDate);
  }

  
  DailyMealPlan? getDailyPlan(DateTime date) {
    final dayOnly = DateTime(date.year, date.month, date.day);
    return dailyPlans.firstWhere(
          (plan) => plan.date.year == dayOnly.year &&
          plan.date.month == dayOnly.month &&
          plan.date.day == dayOnly.day,
      orElse: () => throw Exception('No meal plan for this date'),
    );
  }

  WeeklyMealPlan copyWith({
    int? id,
    DateTime? startDate,
    DateTime? endDate,
    int? userId,
    List<DailyMealPlan>? dailyPlans,
  }) {
    return WeeklyMealPlan(
      id: id ?? this.id,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      userId: userId ?? this.userId,
      dailyPlans: dailyPlans ?? List.from(this.dailyPlans),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is WeeklyMealPlan &&
              runtimeType == other.runtimeType &&
              id == other.id;

  @override
  int get hashCode => id.hashCode;
}

class DailyMealPlan {
  final int id;
  final DateTime date;
  final int caloriesPerDay;
  final int weeklyMealPlanId;

  
  List<Meal> meals = [];
  List<DailyMealPlanIngredient> ingredients = [];

  DailyMealPlan({
    required this.id,
    required this.date,
    required this.caloriesPerDay,
    required this.weeklyMealPlanId,
    List<Meal>? meals,
    List<DailyMealPlanIngredient>? ingredients,
  }) :
        meals = meals ?? [],
        ingredients = ingredients ?? [];

  factory DailyMealPlan.fromJson(Map<String, dynamic> json) {
    return DailyMealPlan(
      id: json['dmp_id'],
      date: DateTime.parse(json['dmp_date']),
      caloriesPerDay: json['dmp_cal_day'] ?? 0,
      weeklyMealPlanId: json['dmp_wmp_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'dmp_id': id,
      'dmp_date': date.toIso8601String().split('T')[0],
      'dmp_cal_day': caloriesPerDay,
      'dmp_wmp_id': weeklyMealPlanId,
    };
  }

  
  int get totalCalories {
    int total = 0;
    for (var meal in meals) {
      for (var recipe in meal.recipes) {
        if (recipe.recipe != null) {
          total += recipe.recipe!.calories;
        }
      }
    }
    return total;
  }

  DailyMealPlan copyWith({
    int? id,
    DateTime? date,
    int? caloriesPerDay,
    int? weeklyMealPlanId,
    List<Meal>? meals,
    List<DailyMealPlanIngredient>? ingredients,
  }) {
    return DailyMealPlan(
      id: id ?? this.id,
      date: date ?? this.date,
      caloriesPerDay: caloriesPerDay ?? this.caloriesPerDay,
      weeklyMealPlanId: weeklyMealPlanId ?? this.weeklyMealPlanId,
      meals: meals ?? List.from(this.meals),
      ingredients: ingredients ?? List.from(this.ingredients),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is DailyMealPlan &&
              runtimeType == other.runtimeType &&
              id == other.id;

  @override
  int get hashCode => id.hashCode;
}

class DailyMealPlanIngredient {
  final int dailyMealPlanId;
  final int ingredientId;
  final int quantity;
  final QuantityType quantityType;

  
  Ingredient? ingredient;

  DailyMealPlanIngredient({
    required this.dailyMealPlanId,
    required this.ingredientId,
    required this.quantity,
    required this.quantityType,
    this.ingredient,
  });

  factory DailyMealPlanIngredient.fromJson(Map<String, dynamic> json) {
    return DailyMealPlanIngredient(
      dailyMealPlanId: json['mid_dmp_id'],
      ingredientId: json['mid_ing_id'],
      quantity: json['mid_quantity'] ?? 0,
      quantityType: QuantityType.fromString(json['mid_quantity_type'] ?? 'grams'),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'mid_dmp_id': dailyMealPlanId,
      'mid_ing_id': ingredientId,
      'mid_quantity': quantity,
      'mid_quantity_type': quantityType.toString().split('.').last,
    };
  }

  
  String get formattedQuantity {
    return '$quantity ${quantityType.getShortName()}';
  }

  DailyMealPlanIngredient copyWith({
    int? dailyMealPlanId,
    int? ingredientId,
    int? quantity,
    QuantityType? quantityType,
    Ingredient? ingredient,
  }) {
    return DailyMealPlanIngredient(
      dailyMealPlanId: dailyMealPlanId ?? this.dailyMealPlanId,
      ingredientId: ingredientId ?? this.ingredientId,
      quantity: quantity ?? this.quantity,
      quantityType: quantityType ?? this.quantityType,
      ingredient: ingredient ?? this.ingredient,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is DailyMealPlanIngredient &&
              runtimeType == other.runtimeType &&
              dailyMealPlanId == other.dailyMealPlanId &&
              ingredientId == other.ingredientId;

  @override
  int get hashCode => Object.hash(dailyMealPlanId, ingredientId);
}

class Meal {
  final int id;
  final DateTime date;
  final int userId;
  final String type;
  final String time;

  
  List<MealRecipe> recipes = [];
  List<MealIngredient> ingredients = [];

  Meal({
    required this.id,
    required this.date,
    required this.userId,
    required this.type,
    required this.time,
    List<MealRecipe>? recipes,
    List<MealIngredient>? ingredients,
  }) :
        recipes = recipes ?? [],
        ingredients = ingredients ?? [];

  factory Meal.fromJson(Map<String, dynamic> json) {
    return Meal(
      id: json['adm_id'],
      date: DateTime.parse(json['adm_date']),
      userId: json['adm_usr_id'],
      type: json['type'] ?? 'Прием пищи',
      time: json['time'] ?? '12:00',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'adm_id': id,
      'adm_date': date.toIso8601String().split('T')[0],
      'adm_usr_id': userId,
      'type': type,
      'time': time,
    };
  }

  
  int get totalCalories {
    int total = 0;
    for (var recipe in recipes) {
      if (recipe.recipe != null) {
        total += recipe.recipe!.calories;
      }
    }
    for (var ingredient in ingredients) {
      if (ingredient.ingredient != null) {
        
        final baseCalories = ingredient.ingredient!.calories;
        final baseWeight = ingredient.ingredient!.weight;
        if (baseWeight > 0 && ingredient.quantity > 0) {
          total += (baseCalories * ingredient.quantity) ~/ baseWeight;
        }
      }
    }
    return total;
  }

  Meal copyWith({
    int? id,
    DateTime? date,
    int? userId,
    String? type,
    String? time,
    List<MealRecipe>? recipes,
    List<MealIngredient>? ingredients,
  }) {
    return Meal(
      id: id ?? this.id,
      date: date ?? this.date,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      time: time ?? this.time,
      recipes: recipes ?? List.from(this.recipes),
      ingredients: ingredients ?? List.from(this.ingredients),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is Meal &&
              runtimeType == other.runtimeType &&
              id == other.id;

  @override
  int get hashCode => id.hashCode;
}

class MealRecipe {
  final int mealId;
  final int recipeId;

  
  Recipe? recipe;

  MealRecipe({
    required this.mealId,
    required this.recipeId,
    this.recipe,
  });

  factory MealRecipe.fromJson(Map<String, dynamic> json) {
    return MealRecipe(
      mealId: json['mra_adm_id'],
      recipeId: json['mra_rcp_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'mra_adm_id': mealId,
      'mra_rcp_id': recipeId,
    };
  }

  MealRecipe copyWith({
    int? mealId,
    int? recipeId,
    Recipe? recipe,
  }) {
    return MealRecipe(
      mealId: mealId ?? this.mealId,
      recipeId: recipeId ?? this.recipeId,
      recipe: recipe ?? this.recipe,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is MealRecipe &&
              runtimeType == other.runtimeType &&
              mealId == other.mealId &&
              recipeId == other.recipeId;

  @override
  int get hashCode => Object.hash(mealId, recipeId);
}

class MealIngredient {
  final int mealId;
  final int ingredientId;
  final int quantity;
  final QuantityType quantityType;

  
  Ingredient? ingredient;

  MealIngredient({
    required this.mealId,
    required this.ingredientId,
    required this.quantity,
    required this.quantityType,
    this.ingredient,
  });

  factory MealIngredient.fromJson(Map<String, dynamic> json) {
    return MealIngredient(
      mealId: json['mia_adm_id'],
      ingredientId: json['mia_ing_id'],
      quantity: json['mia_quantity'] ?? 0,
      quantityType: QuantityType.fromString(json['mia_quantity_type'] ?? 'grams'),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'mia_adm_id': mealId,
      'mia_ing_id': ingredientId,
      'mia_quantity': quantity,
      'mia_quantity_type': quantityType.toString().split('.').last,
    };
  }

  
  String get formattedQuantity {
    return '$quantity ${quantityType.getShortName()}';
  }

  MealIngredient copyWith({
    int? mealId,
    int? ingredientId,
    int? quantity,
    QuantityType? quantityType,
    Ingredient? ingredient,
  }) {
    return MealIngredient(
      mealId: mealId ?? this.mealId,
      ingredientId: ingredientId ?? this.ingredientId,
      quantity: quantity ?? this.quantity,
      quantityType: quantityType ?? this.quantityType,
      ingredient: ingredient ?? this.ingredient,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is MealIngredient &&
              runtimeType == other.runtimeType &&
              mealId == other.mealId &&
              ingredientId == other.ingredientId;

  @override
  int get hashCode => Object.hash(mealId, ingredientId);
}

File: AppBackend/apps/core/models/meal_plan.py

from django.db import models
from django.contrib.auth import get_user_model
from django.utils.translation import gettext_lazy as _
from ..enums import QuantityTypeChoices

User = get_user_model()

class WeaklyMealPlan(models.Model):
    """Модель недельного плана питания"""
    wmp_id = models.AutoField(primary_key=True)
    wmp_start = models.DateField(_('Дата начала'))
    wmp_end = models.DateField(_('Дата окончания'))
    wmp_usr_id = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        db_column='wmp_usr_id',
        related_name='weekly_meal_plans',
        verbose_name=_('Пользователь')
    )

    class Meta:
        db_table = 'weakly_meal_plan'
        verbose_name = _('Недельный план питания')
        verbose_name_plural = _('Недельные планы питания')

    def __str__(self):
        return f"План {self.wmp_id} для {self.wmp_usr_id} ({self.wmp_start} - {self.wmp_end})"


class DailyMealPlan(models.Model):
    """Модель дневного плана питания"""
    dmp_id = models.AutoField(primary_key=True)
    dmp_date = models.DateField(_('Дата'))
    dmp_cal_day = models.IntegerField(_('Дневная норма калорий'), default=0)
    dmp_wmp_id = models.ForeignKey(
        'WeaklyMealPlan',
        on_delete=models.CASCADE,
        db_column='dmp_wmp_id',
        related_name='daily_plans',
        verbose_name=_('Недельный план')
    )

    class Meta:
        db_table = 'daily_meal_plan'
        verbose_name = _('Дневной план питания')
        verbose_name_plural = _('Дневные планы питания')

    def __str__(self):
        return f"План на {self.dmp_date}"


class ActualDayMeal(models.Model):
    """Модель фактического приема пищи"""
    adm_id = models.AutoField(primary_key=True)
    adm_date = models.DateField(_('Дата'))
    adm_usr_id = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        db_column='adm_usr_id',
        related_name='actual_meals',
        verbose_name=_('Пользователь')
    )
    
    adm_type = models.CharField(_('Тип приема пищи'), max_length=50, null=True, blank=True)  
    adm_time = models.TimeField(_('Время'), null=True, blank=True)

    class Meta:
        db_table = 'actual_day_meal'
        verbose_name = _('Фактический прием пищи')
        verbose_name_plural = _('Фактические приемы пищи')

    def __str__(self):
        meal_type = f" ({self.adm_type})" if self.adm_type else ""
        return f"Прием пищи на {self.adm_date}{meal_type}"


class M2MRcpDmp(models.Model):
    """Связь между рецептом и дневным планом питания"""
    mrd_id = models.AutoField(primary_key=True)
    mrd_rcp_id = models.ForeignKey(
        'Recipe',
        on_delete=models.CASCADE,
        db_column='mrd_rcp_id',
        related_name='daily_meal_plans',
        verbose_name=_('Рецепт')
    )
    mrd_dmp_id = models.ForeignKey(
        'DailyMealPlan',
        on_delete=models.CASCADE,
        db_column='mrd_dmp_id',
        related_name='recipes',
        verbose_name=_('Дневной план')
    )

    class Meta:
        db_table = 'm2m_rcp_dmp'
        verbose_name = _('Рецепт в дневном плане')
        verbose_name_plural = _('Рецепты в дневных планах')


class M2MRcpAdm(models.Model):
    """Связь между рецептом и приемом пищи"""
    mra_adm_id = models.ForeignKey(
        'ActualDayMeal',
        on_delete=models.CASCADE,
        db_column='mra_adm_id',
        related_name='recipes',
        verbose_name=_('Прием пищи')
    )
    mra_rcp_id = models.ForeignKey(
        'Recipe',
        on_delete=models.CASCADE,
        db_column='mra_rcp_id',
        related_name='actual_meals',
        verbose_name=_('Рецепт')
    )

    class Meta:
        db_table = 'm2m_rcp_adm'
        unique_together = ('mra_adm_id', 'mra_rcp_id')
        verbose_name = _('Рецепт в приеме пищи')
        verbose_name_plural = _('Рецепты в приемах пищи')


class M2MIngDmp(models.Model):
    """Связь между ингредиентом и дневным планом питания"""
    mid_dmp_id = models.ForeignKey(
        'DailyMealPlan',
        on_delete=models.CASCADE,
        db_column='mid_dmp_id',
        related_name='ingredients',
        verbose_name=_('Дневной план')
    )
    mid_ing_id = models.ForeignKey(
        'Ingredient',
        on_delete=models.CASCADE,
        db_column='mid_ing_id',
        related_name='daily_plans',
        verbose_name=_('Ингредиент')
    )
    mid_quantity = models.IntegerField(_('Количество'), default=0)
    mid_quantity_type = models.CharField(
        _('Единица измерения'),
        max_length=20,
        choices=QuantityTypeChoices.choices,
        default=QuantityTypeChoices.GRAMS
    )

    class Meta:
        db_table = 'm2m_ing_dmp'
        unique_together = ('mid_dmp_id', 'mid_ing_id')
        verbose_name = _('Ингредиент в дневном плане')
        verbose_name_plural = _('Ингредиенты в дневных планах')


class M2MIngAdm(models.Model):
    """Связь между ингредиентом и приемом пищи"""
    mia_adm_id = models.ForeignKey(
        'ActualDayMeal',
        on_delete=models.CASCADE,
        db_column='mia_adm_id',
        related_name='ingredients',
        verbose_name=_('Прием пищи')
    )
    mia_ing_id = models.ForeignKey(
        'Ingredient',
        on_delete=models.CASCADE,
        db_column='mia_ing_id',
        related_name='actual_meals',
        verbose_name=_('Ингредиент')
    )
    mia_quantity = models.IntegerField(_('Количество'), default=0)
    mia_quantity_type = models.CharField(
        _('Единица измерения'),
        max_length=20,
        choices=QuantityTypeChoices.choices,
        default=QuantityTypeChoices.GRAMS
    )

    class Meta:
        db_table = 'm2m_ing_adm'
        unique_together = ('mia_adm_id', 'mia_ing_id')
        verbose_name = _('Ингредиент в приеме пищи')
        verbose_name_plural = _('Ингредиенты в приемах пищи')